Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp

```diff
@@ -1128,6 +1129,69 @@ void MediaPlayerPrivateGStreamer::notifyPlayerOfVideoCaps()
     m_player->mediaEngineUpdated();
 }
 
+String audioTrackIdFromStreamStart(GstEvent* streamStart) {
+    const gchar* streamIdAsCharacters;
+    gst_event_parse_stream_start(streamStart, &streamIdAsCharacters);
+
+    String streamId;
+    if (streamIdAsCharacters) {
+        StringView streamIdView(streamIdAsCharacters);
+        size_t position = streamIdView.find('/');
+
+        if (position != notFound && position + 1 < streamIdView.length())
+            streamId = streamIdView.substring(position + 1).toString();
+    }
+
+    return streamId;
+}
+
+class StreamStartProbeData : public ThreadSafeRefCounted<StreamStartProbeData> {
+public:
+    StreamStartProbeData(MediaPlayerPrivateGStreamer* player, int trackIndex)
+        : player(player)
+        , trackIndex(trackIndex)
+    { }
+
+    MediaPlayerPrivateGStreamer* player;
+    int trackIndex;
+    bool probeWasRemoved { false };
+    Lock lock;
+};
+
+GstPadProbeReturn MediaPlayerPrivateGStreamer::streamStartProbeCallback(GstPad*, GstPadProbeInfo* info, gpointer data)
+{
+    Ref<StreamStartProbeData> probeData(*static_cast<StreamStartProbeData*>(data));
+    LockHolder locker(probeData->lock);
+    GstEvent* event = GST_PAD_PROBE_INFO_EVENT(info);
+
+    if (UNLIKELY(probeData->probeWasRemoved) || GST_EVENT_TYPE(event) != GST_EVENT_STREAM_START)
```

Comment:
I would split this condition into two: first I'd check the stream start type and bail out and then, get the probeData, lock and check for the probe to be removed.